src/shared/components/base-question-app.tsx

```diff
@@ -0,0 +1,85 @@
+import React, { useRef } from "react";
+import { useLARAInteractiveAPI } from "../hooks/use-lara-interactive-api";
+import { useAutoHeight } from "../hooks/use-auto-height";
+import { useHint } from "../hooks/use-hint";
+import { useRequiredQuestion } from "../hooks/use-required-question";
+import { BaseAuthoring, IBaseAuthoringProps } from "./base-authoring";
+
+import css from "./base-question-app.scss";
+
+interface IBaseQuestionAuthoredState {
+  version: number;
+  hint?: string;
+  required?: boolean;
+}
+
+interface IBaseQuestionInteractiveState {
+  submitted?: boolean;
+}
+
+interface IAuthoringComponentProps<IAuthoredState> {
+  authoredState: IAuthoredState | undefined,
+  setAuthoredState?: (state: IAuthoredState) => void;
+}
+
+interface IRuntimeComponentProps<IAuthoredState, IInteractiveState> {
+  authoredState: IAuthoredState,
+  interactiveState: IInteractiveState | undefined,
+  setInteractiveState?: (state: IInteractiveState) => void;
+  report?: boolean;
+}
+
+interface IProps<IAuthoredState, IInteractiveState> {
+  // Question can provide either fully custom Authoring component or use BaseAuthoring (that uses react-jsonschema-form).
+  Authoring?: React.FC<IAuthoringComponentProps<IAuthoredState>>;
+  baseAuthoringProps?: Omit<IBaseAuthoringProps<IAuthoredState>, "authoredState" | "setAuthoredState">;
+  Runtime: React.FC<IRuntimeComponentProps<IAuthoredState, IInteractiveState>>;
+  isAnswered: (state: IInteractiveState | undefined) => boolean;
+}
+
+export const BaseQuestionApp = <IAuthoredState extends IBaseQuestionAuthoredState, IInteractiveState extends IBaseQuestionInteractiveState>(
+  { Authoring, baseAuthoringProps, Runtime, isAnswered }: IProps<IAuthoredState, IInteractiveState>
+) => {
+  const container = useRef<HTMLDivElement>(null);
+  const { mode, authoredState, interactiveState, setInteractiveState, setAuthoredState, setHeight, setHint, setNavigation } = useLARAInteractiveAPI<IAuthoredState, IInteractiveState>({
+    interactiveState: true,
+    authoredState: true,
+  });
+  useAutoHeight({ container, setHeight });
+  useHint({ authoredState, setHint });
+  const { submitButton, lockedInfo } = useRequiredQuestion({ authoredState, interactiveState, setInteractiveState, setNavigation, isAnswered: isAnswered(interactiveState) });
+
+  const report = mode === "report";
+  const authoring = mode === "authoring";
+  const runtime = mode === "runtime";
+
+  let AuthoringComp = null;
+  if (authoring) {
+    if (Authoring) {
```

Comment:
This use of local variable names that differ only in case confused me initially. Consider `reportMode`, `authoringMode`, `runtimeMode` or some other such renaming to make the distinction more readily apparent.